# Download handbook and data  


## Download offline handbook  

You can download the offline version of this handbook as an HTML file. Once downloaded, you can view the file in your web browser even if you no longer have internet access.  

Note: The file is large (>40MB). When you open it, it may take several minutes for the images and the Table of Contents to load. Also, the offline handbook displays as one long page, so please search for specific content using with Ctrl+f (Cmd-f). 
If you are considering offline use, please also see the page [Suggested packages] to assist you with installing appropriate R packages before you lose internet connectivity.  


**There are two ways you can download the handbook:**  

### Option 1: Download link {-}  

Right-click this [link](https://github.com/epirhandbook/Epi_R_handbook/raw/master/offline_long/Epi_R_Handbook_offline.html) and select "Save link as".  

If on a Mac, use Cmd+click. If on a mobile, press and hold the link and select "Save link". The handbook will download to your device. If a screen with raw HTML code appears, ensure you follow the above instructions.  


### Option 2: Download via our R package {-}  

Alternatively, you can download the handbook using our R package **handbookfiles**.  

Run the following code to install our R package **handbookfiles** from the Github repository *epirhandbook*. This package is not on CRAN, so use the special function `p_install_gh()` to install it from Github.  

```{r, eval=F}
pacman::p_install_gh("epirhandbook/handbookfiles")
```

Now, run the function `download_book()` (with empty parentheses) to download the handbook to your computer. By default a pop-up window will appear, allowing you to select a save location. Alternatively, you can provide a file path to `save_loc = `.  

```{r, eval=F}
handbookfiles::download_book()
```



## Download data to follow along  

To "follow along" with the pages, you can download the datasets and use them yourself. See the instructions for download below. All the data in this handbook are available from the **["data" folder](https://github.com/nsbatra/Epi_R_handbook/tree/master/data)** of our Github repository.  





### Case linelist {-}

This is a fictional Ebola outbreak, expanded by the handbook team from the `ebola_sim` practice dataset in the **outbreaks** package.  

* <a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/linelist_raw.xlsx' class='download-button'>Click to download the "raw" linelist (.xlsx)</span></a>. The "raw" case linelist is an Excel spreadsheet with messy data. Use this to follow-along with the [Cleaning data and core functions] page.  

* <a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/linelist_cleaned.rds' class='download-button'>Click to download the "clean" linelist (.rds)</a>. Use this file for all other pages of this handbook that use the linelist. A .rds file is an R-specific file type that preserves column classes. This ensures you will have only minimal cleaning to do after importing the data into R.  

*Other related files:*  

* <a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/linelist_cleaned.xlsx' class='download-button'>Click to download the "clean" linelist as an Excel file</a>

* Part of the cleaning page uses a "cleaning dictionary" (.csv file). You can load it directly into R by running the following commands in R.  

```{r, eval=F}
pacman::p_load(rio) # install/load the rio package

# import the file directly from Github
cleaning_dict <- import("https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/cleaning_dict.csv")
```


### Malaria count data {-}  

These data are fictional counts of malaria cases by age group, facility, and day. A .rds file is an R-specific file type that preserves column classes. This ensures you will have only minimal cleaning to do after importing the data into R.  

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/malaria_facility_count_data.rds' class='download-button'>
	Click to download
	<span>the malaria count data (.rds file)</span>
</a>


### Likert-scale data {-}  

These are fictional data from a Likert-style survey, used in the page on [Demographic pyramids and Likert-scales]. You can load these data directly into R by running the following commands in R.  

```{r, eval=F}
pacman::p_load(rio) # install/load the rio package

# import the file directly from Github
likert_data <- import("https://raw.githubusercontent.com/nsbatra/Epi_R_handbook/master/data/likert_data.csv")
```


### Flexdashboard {-}  

Below are links to the file associated with the page on [Dashboards with R Markdown]:  

* To download the R Markdown for the outbreak dashboard, right-click this [link](https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/flexdashboard/outbreak_dashboard.Rmd) (Cmd+click for Mac) and select "Save link as".  
* To download the HTML dashboard, right-click this [link](https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/flexdashboard/outbreak_dashboard_test.html) (Cmd+click for Mac) and select "Save link as".  


### GIS {-}  

UNDER CONSTRUCTION

Shapefiles have many sub-component files, each with a different file extention. One file will have the ".shp" extension, but others may have ".dbf", ".prj", etc.  

The [GIS basics] page provides links to the *Humanitarian Data Exchange* website where you can download the shapefiles directly. For example, the health facility points can be downloaded [here](https://data.humdata.org/dataset/hotosm_sierra_leone_health_facilities). Download the points shapefile (shp) ".zip" folder. Once saved to your computer, "unzip" the folder. You will see several files with different extensions (e.g. ".shp", ".prj", ".shx") - all these must be saved in the same folder. As described in the [GIS basics] page, provide the filepath and name of the ".shp" file to `st_read()` from the **sf** package.  

Alternatively, you can download the shapefiles from the R Handbook Github "data" folder (see the "gis" sub-folder). However, be aware that you will need to download *each* sub-file individually to your computer. In Github, click on each file individually and download them by clicking on the "Download" button. Below, you can see how the shapefile "sle_adm3" consists of many files - each of which must be downloaded from Github.  

```{r out.height = "50%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_shp.png"))
```


### Phylogenetic trees {-}  

See the page on [Phylogenetic trees]. Newick file of phylogenetic tree constructed from whole genome sequencing of 299 Shigella sonnei samples and corresponding sample data (converted to a text file). The Belgian samples and resulting data are kindly provided by the Belgian NRC for Salmonella and Shigella in the scope of a project conducted by an ECDC EUPHEM Fellow, and will also be published in a manuscript. The international data are openly available on public databases (ncbi) and have been previously published.  

* To download the "Shigella_tree.txt" phylogenetic tree file, right-click this [link](https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/phylo/Shigella_tree.txt) (Cmd+click for Mac) and select "Save link as".  
* To download the "sample_data_Shigella_tree.csv" with additional information on each sample, right-click this [link](https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/phylo/sample_data_Shigella_tree.csv) (Cmd+click for Mac) and select "Save link as".  
* To see the new, created subset-tree, right-click this [link](https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/phylo/Shigella_subtree_2.txt) (Cmd+click for Mac) and select "Save link as". The .txt file will download to your computer.  


You can then import the .txt files with `read.tree()` from the **ape** package, as explained in the page.

```{r, eval=F}
ape::read.tree("Shigella_tree.txt")
```


### Standardization {-}  

See the page on [Standardization]. You can load the data directly from our Github repository on the internet into your R session with the following commands:  


```{r, eval=F}
# install/load the rio package
pacman::p_load(rio) 

##############
# Country A
##############
# import demographics for country A directly from Github
countryAdemo <- import("https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/country_demographics.csv")

# import deaths for country A directly from Github
A_deaths <- import("https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/deaths_countryA.csv")

##############
# Country B
##############
# import demographics for country B directly from Github
countryBdemo <- import("https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/country_demographics_2.csv")

# import deaths for country B directly from Github
B_deaths <- import("https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/deaths_countryB.csv")


###############
# Reference Pop
###############
# import demographics for country B directly from Github
standard_pop_data <- import("https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/world_standard_population_by_sex.csv")
```



### Time series and outbreak detection {-}  

See the page on [Time series and outbreak detection]. We use campylobacter cases reported in Germany 2002-2011, as available from the **surveillance** R package.

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/campylobacter_germany.xlsx' class='download-button'>
	Click to download
	<span> Campylobacter in Germany (.xlsx)</span>
</a>

Also used are climate data (temperature in degrees celsius and rain fail in millimetres) in Germany 2002-2011. These were downloaded from the EU Copernicus satellite reanalysis dataset using the **ecmwfr** package. You will need to download all of these and import them with `stars::read_stars()` as explained in the time series page.  

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2002.nc' class='download-button'>
	Click to download
	<span> Germany weather 2002 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2003.nc' class='download-button'>
	Click to download
	<span> Germany weather 2003 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2004.nc' class='download-button'>
	Click to download
	<span> Germany weather 2004 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2005.nc' class='download-button'>
	Click to download
	<span> Germany weather 2005 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2006.nc' class='download-button'>
	Click to download
	<span> Germany weather 2006 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2007.nc' class='download-button'>
	Click to download
	<span> Germany weather 2007 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2008.nc' class='download-button'>
	Click to download
	<span> Germany weather 2008 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2009.nc' class='download-button'>
	Click to download
	<span> Germany weather 2009 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2010.nc' class='download-button'>
	Click to download
	<span> Germany weather 2010 (.nc file)</span>
</a> 

<a href='https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2011.nc' class='download-button'>
	Click to download
	<span> Germany weather 2011 (.nc file)</span>
</a>


